### PR TITLE
[FIX] module logical: vectorize only if valid expression

### PR DESCRIPTION
**Pull request by laa-odoo**
In logical formulas, avoid vectorizing arguments not affected by the logical expression.

Task: [4775222](https://www.odoo.com/odoo/2328/tasks/4775222)

### CODE REVIEW

File: src/functions/index.ts [hunk 1]
```diff
@@ -277,6 +207,86 @@ export function handleError(e: unknown, functionName: string): FunctionResultObj
   );
 }
 
+export function applyVectorization<T extends Arg[]>(
```
Comment (VincentSchippefilt) [inline src/functions/index.ts:210]: Document me, sempaï...
Who should call me, why and when ?

File: src/functions/helper_matrices.ts [hunk 2]
```diff
@@ -139,12 +139,26 @@ export function multiplyMatrices(matrix1: Matrix<number>, matrix2: Matrix<number
...
+}
+
+function isOneOneMatrix<T>(matrix: Matrix<T>) {
+  if (matrix.length !== 1) {
+    return false;
+  }
+  if (matrix[0].length !== 1) {
+    return false;
+  }
+  return true;
```
Comment (hokolomopo) [inline src/functions/helper_matrices.ts:159]:
```suggestion
    return matrix.length === 1 && matrix[0].length === 1;
```

Comment (hokolomopo) [reply src/functions/helper_matrices.ts:159]: And rename to `isOneByOneMatrix` matrix ? `isSingleCellMatrix` ?

File: src/functions/module_logical.ts [hunk 3]
```diff
@@ -49,27 +51,33 @@ export const IF = {
...
-      "logical_expression (boolean)",
+      "logical_expression (boolean, range<boolean>)",
       _t(
         "An expression or reference to a cell containing an expression that represents some logical value, i.e. TRUE or FALSE."
       )
     ),
-    arg("value_if_true (any)", _t("The value the function returns if logical_expression is TRUE.")),
     arg(
-      "value_if_false (any, default=FALSE)",
+      "value_if_true (any, range)",
```
Comment (hokolomopo) [inline src/functions/module_logical.ts:60]:
```suggestion
      "value_if_true (any, range<any>)",
```
probably a bit more clear no ? Otherwise I'm wondering why `range` is not included in `any`

**Review by hokolomopo — COMMENTED** [review]
I didn't follow the discussion at all, so it was just a fast scroll scroll and I'll let someone that know what you did reveiw it correctly :)

File: src/functions/helper_matrices.ts (hunk 2 above)
Comment (laa-odoo) [reply src/functions/helper_matrices.ts:159]: thx

File: src/functions/module_logical.ts (hunk 3 above)
Comment (laa-odoo) [reply src/functions/module_logical.ts:60]: both notations are possible, we should clean everything up if we want only one representation

Comment (rrahir) [reply src/functions/module_logical.ts:65]: if both are possible (atm) i think it doesn't hurt to add <any> to avoid misunderstandings

File: src/functions/module_logical.ts [hunk 4]
```diff
@@ -49,27 +56,33 @@ export const IF = {
...
-    valueIfTrue: Maybe<FunctionResultObject>,
-    valueIfFalse: Maybe<FunctionResultObject>
-  ): FunctionResultObject {
-    const result = toBoolean(logicalExpression?.value) ? valueIfTrue : valueIfFalse;
+  compute: function (logicalExpression: Arg, valueIfTrue: Arg, valueIfFalse: Arg) {
+    if (isMultipleElementMatrix(logicalExpression)) {
+      return applyVectorization(IF.compute, [logicalExpression, valueIfTrue, valueIfFalse]);
+    }
+    let result = toBoolean(toScalar(logicalExpression)) ? valueIfTrue : valueIfFalse;
+    // useful for interpreting empty cell references as empty strings. But must be removed to make empty cell references equal to zero
```
Comment (rrahir) [inline src/functions/module_logical.ts:78]: Could you add a dedicated test to this case so we can state clearly "we have decided to return empty stringsin this specific case" ?

**Comment by rrahir** [comment]
@robodoo r+

**Comment by laa-odoo** [comment]
rebase-ff

**Comment by laa-odoo** [comment]
robodoo rebase-ff
